### PR TITLE
Fix/2257 filter entity by relation

### PR DESCRIPTION
**Pull request by abrl91**
#2257 

Added the ability to filter an entity by relation, for example: 
all the users with posts that have less than 50 likes

```
prisma.user.findMany({
  where: {
    posts: {
      some: {
        likes: {
          lte: 50,
        },
      },
    },
  },
})
```

@yuval-hazaz 
I made some changes that I am not sure are needed:
- to-many.template.ts
- create-resolver.ts
- create-controller.ts

### CODE REVIEW

File: packages/amplication-data-service-generator/src/server/resource/create-dtos.ts [hunk 1]
```diff
@@ -31,6 +32,7 @@ type EntityDTOs = {
   createArgs?: NamedClassDeclaration;
   updateArgs?: NamedClassDeclaration;
   orderByInput: NamedClassDeclaration;
+  entityListRelationFilter: NamedClassDeclaration;
```
Comment (yuval-hazaz) [inline packages/amplication-data-service-generator/src/server/resource/create-dtos.ts:35]: rename to `listRelationFilter`

File: packages/amplication-data-service-generator/src/server/resource/dto/create-field-class-property.ts [hunk 2]
```diff
@@ -266,7 +272,9 @@ export function createFieldClassProperty(
         TYPE_ID,
         builders.arrowFunctionExpression(
           [],
-          prismaField.isList ? builders.arrayExpression([typeName]) : typeName
+          prismaField.isList && isToManyRelationField(field) && !isQuery
```
Comment (yuval-hazaz) [inline packages/amplication-data-service-generator/src/server/resource/dto/create-field-class-property.ts:275]: This seems incorrect - see the comment on the snapshot about "favoriteColors" - I think this is the cause for the unexpected behavior

File: packages/amplication-data-service-generator/src/tests/__snapshots__/create-data-service.spec.ts.snap [hunk 3]
```diff
@@ -3456,7 +3505,7 @@ class Customer {
     each: true,
   })
   @IsOptional()
-  @Field(() => [EnumCustomerFavoriteColors], {
+  @Field(() => EnumCustomerFavoriteColors, {
```
Comment (yuval-hazaz) [inline packages/amplication-data-service-generator/src/tests/__snapshots__/create-data-service.spec.ts.snap:3508]: This should not change
See the comment in packages/amplication-data-service-generator/src/server/resource/dto/create-field-class-property.ts

File: packages/amplication-data-service-generator/src/server/resource/dto/create-field-class-property.ts [hunk 4]
```diff
@@ -340,7 +348,7 @@ function createGraphQLFieldType(
   entity: Entity,
   isQuery: boolean
 ): namedTypes.Identifier | namedTypes.ArrayExpression {
-  if (prismaField.isList) {
+  if (prismaField.isList && isToManyRelationField(field) && !isQuery) {
```
Comment (yuval-hazaz) [inline packages/amplication-data-service-generator/src/server/resource/dto/create-field-class-property.ts:351]: This is incorrect, the change incorrectly exclude fields when isList && !isRelation

**Review by yuval-hazaz — CHANGES_REQUESTED** [review]
@abrl91 impressive work here! well done

Please see my comments about the conditions 
Once we fix it, I will review again all other parts

File: packages/amplication-data-service-generator/src/server/resource/dto/create-field-class-property.ts (hunk 4 above)
Comment (abrl91) [reply packages/amplication-data-service-generator/src/server/resource/dto/create-field-class-property.ts:351]: we need that when it is prisma field list and to-many relation and not query that the type would not be inside an array,
otherwise, this condition is true. Then, on line 390 we have the following condition:

```
if (isToManyRelationField(field)) {
    return createEntityListRelationFilterID(prismaField.type);
  }
```

And as a result we got [EntityListRelationField]

File: packages/amplication-data-service-generator/src/tests/__snapshots__/create-data-service.spec.ts.snap (hunk 3 above)
Comment (abrl91) [reply packages/amplication-data-service-generator/src/tests/__snapshots__/create-data-service.spec.ts.snap:3508]: I think I fixed it. I hope it didn't break something else. Every time I change a condition, it breaks another part.
If nobody wants to refactor this file, I would love to :)

File: packages/amplication-data-service-generator/src/server/resource/dto/create-field-class-property.ts (hunk 2 above)
Comment (abrl91) [reply packages/amplication-data-service-generator/src/server/resource/dto/create-field-class-property.ts:275]: this block code determines the type of the 'ApiProperty'

File: packages/amplication-data-service-generator/src/server/resource/dto/graphql/entity-list-relation-filter/create-entity-list-relation.spec.ts [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+describe("createEntityListRelation", () => {
+  test("create list relation", async () => {
+    expect(
+      print(
+        await createEntityListRelationFilter(
+          EXAMPLE_ENTITY,
+          EXAMPLE_WHERE_INPUT
+        )
+      ).code
+    ).toEqual(`@InputType()
```
Comment (tupe12334) [inline packages/amplication-data-service-generator/src/server/resource/dto/graphql/entity-list-relation-filter/create-entity-list-relation.spec.ts:32]: Way not to use snapshot inline test instead?

Comment (tupe12334) [reply packages/amplication-data-service-generator/src/server/resource/dto/graphql/entity-list-relation-filter/create-entity-list-relation.spec.ts:32]: https://jestjs.io/docs/snapshot-testing#inline-snapshots

Comment (abrl91) [reply packages/amplication-data-service-generator/src/server/resource/dto/graphql/entity-list-relation-filter/create-entity-list-relation.spec.ts:32]: I saw all the other Input has test files, so I added one too

Comment (tupe12334) [reply packages/amplication-data-service-generator/src/server/resource/dto/graphql/entity-list-relation-filter/create-entity-list-relation.spec.ts:32]: Yes, all the other inputs have a test that tests there AST object.
You made a good test (my preferred way ;) ) that checked the generation of the dto, i just suggest you a better way (in my opinion) to do a string compressing test, a snapshot test

Comment (abrl91) [reply packages/amplication-data-service-generator/src/server/resource/dto/graphql/entity-list-relation-filter/create-entity-list-relation.spec.ts:32]: I am not sure why they say on the docs:
"Snapshot tests are a very useful tool whenever you want to make sure your UI does not change unexpectedly."
A template creates EntityListRelationFilter, and it has no reason to change unless someone changes the template.
Correct me if I am wrong

Comment (abrl91) [reply packages/amplication-data-service-generator/src/server/resource/dto/graphql/entity-list-relation-filter/create-entity-list-relation.spec.ts]: I changed it to snapshot test. Thanks :)

File: packages/amplication-data-service-generator/src/server/seed/seed.template.ts [hunk 6]
```diff
@@ -41,7 +41,7 @@ async function seed(bcryptSalt: Salt) {
     update: {},
     create: data,
   });
-  client.$disconnect();
+  void client.$disconnect();
```
Comment (yuval-hazaz) [inline packages/amplication-data-service-generator/src/server/seed/seed.template.ts:44]: why did you add this?

Comment (abrl91) [reply packages/amplication-data-service-generator/src/server/seed/seed.template.ts:44]: the lint screamed at me and asked me to add this 😅
look on [17004e5](https://github.com/amplication/amplication/pull/2295/commits/17004e528fad61d6124bdb5ce4c72facedeea14c)

**Review by yuval-hazaz — APPROVED** [review]
Great work @abrl91
